pkg/runtime/runtime_test.go

```diff
@@ -5392,3 +5371,201 @@ func TestNamespacedPublisher(t *testing.T) {
 
 	assert.Equal(t, "ns1topic0", rt.pubSubs[TestPubsubName].component.(*mockPublishPubSub).PublishedRequest.Topic)
 }
+
+func TestGracefulShutdownPubSub(t *testing.T) {
+	rt := NewTestDaprRuntime(modes.StandaloneMode)
+	// setup pubsub
+	testGracefulRestartPubSub := "shutdownPubsub"
+	pubsubComponent := componentsV1alpha1.Component{
+		ObjectMeta: metaV1.ObjectMeta{
+			Name: testGracefulRestartPubSub,
+		},
+		Spec: componentsV1alpha1.ComponentSpec{
+			Type:     "pubsub.mockPubSub",
+			Version:  "v1",
+			Metadata: getFakeMetadataItems(),
+		},
+	}
+	rt.pubSubRegistry.RegisterComponent(
+		func(_ logger.Logger) pubsub.PubSub {
+			return &mockSubscribePubSub{}
+		},
+		"mockPubSub",
+	)
+
+	req := invokev1.NewInvokeMethodRequest("dapr/subscribe")
+	req.WithHTTPExtension(http.MethodGet, "")
+	req.WithRawData(nil, invokev1.JSONContentType)
+	subscriptionItems := []runtimePubsub.SubscriptionJSON{
+		{PubsubName: testGracefulRestartPubSub, Topic: "topic0", Route: "shutdown"},
+	}
+	sub, _ := json.Marshal(subscriptionItems)
+	fakeResp := invokev1.NewInvokeMethodResponse(200, "OK", nil)
+	fakeResp.WithRawData(sub, "application/json")
+
+	mockAppChannel := new(channelt.MockAppChannel)
+	rt.appChannel = mockAppChannel
+	mockAppChannel.On("InvokeMethod", mock.MatchedBy(matchContextInterface), req).Return(fakeResp, nil)
+
+	require.NoError(t, rt.initPubSub(pubsubComponent))
+	rt.startSubscriptions()
+	assert.NotNil(t, rt.pubsubCtx)
+	assert.NotNil(t, rt.topicCtxCancels)
+	assert.NotNil(t, rt.topicRoutes)
+
+	rt.runtimeConfig.GracefulShutdownDuration = 3 * time.Second
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		<-sigs
+		rt.Shutdown(rt.runtimeConfig.GracefulShutdownDuration)
+	}()
+
+	if p, err := os.FindProcess(os.Getpid()); err != nil {
+		rt.Shutdown(rt.runtimeConfig.GracefulShutdownDuration)
+	} else {
+		p.Signal(syscall.SIGTERM)
+	}
+	time.Sleep(1 * time.Second)
+	assert.Nil(t, rt.pubsubCtx)
```

Comment:
instead of sleeping for 1s, can this be blocked on something like this :
```golang
select {
  case <-rt.pubsubCtx.Done():
     // assert context closed, topic context closed, topic routes nil 
    // assert elapsed time is less than graceful shutdown time
  case <-time.After(gracefulShutdownTime):
    // fail the test as this case should not be triggered    t.Fail()
}

```

Reply:
We would need to check the cleanup after graceperiod expires. Updated as suggested

Reply:
```suggestion
	<-time.After(rt.runtimeConfig.GracefulShutdownDuration + 2*time.Second)
	assert.Nil(t, rt.pubsubCtx)
	assert.Nil(t, rt.topicCtxCancels)
	assert.Nil(t, rt.topicRoutes)
```

IMO the main logic that needs to be tested here is that Pubsub subscriptions and input bindings  are closed before the `GracefulShutdownDuration`. But the assertions only happen 2 seconds after graceful shutdown period. 

The logic that has been changed is the shut down sequence and the order of operations matter.
Similarly before the graceful shut down, we can verify that the api closers are not called.